File: frontend/src/lib/receipt.ts
/**
 * Receipt verification, done in the browser.
 *
 * The point of a receipt is that you do not have to take our word for it, so this recomputes every
 * hash locally with the Web Crypto API rather than asking the server whether it is honest.
 *
 * A stored receipt is canonical JSON: keys sorted, no whitespace. Its hash was taken over exactly
 * that text with the `hash` field removed. Rather than re-serialising the parsed object — where a
 * float that Python writes as `1.0` and JavaScript writes as `1` would produce a false mismatch —
 * we cut the `"hash":"..."` pair out of the stored line and hash the remaining text. That is byte
 * for byte what the writer hashed.
 */

import { getToken } from './api'

export const GENESIS = '0'.repeat(64)

export interface Receipt {
  ledger_version: number
  created_utc: string
  capture: { sha256: string; samples?: number; fs_hz?: number | null }
  engine: Record<string, unknown>
  decision: { pack_id?: string | null; status?: string | null; code?: string | null; payload_bits?: number }
  statistics: { log10_p?: number | null; log10_threshold?: number | null; n_hypotheses?: number | null }
  source: Record<string, unknown> | null
  prev_hash: string
  hash: string
}

export interface ChainResult {
  ok: boolean
  checked: number
  head: string
  problems: { index: number; receipt: string; problem: string }[]
}

const HASH_FIELD = /"hash":"[0-9a-f]{64}"/

/** The text a receipt's hash was taken over: the stored line without its own `hash` field. */
export function bodyText(line: string): string | null {
  const m = HASH_FIELD.exec(line)
  if (!m) return null
  const start = m.index, end = m.index + m[0].length
  // Remove the pair and exactly one adjacent comma, whichever side it sits on.
  if (line[start - 1] === ',') return line.slice(0, start - 1) + line.slice(end)
  if (line[end] === ',') return line.slice(0, start) + line.slice(end + 1)
  return line.slice(0, start) + line.slice(end)
}

export async function sha256Hex(text: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text))
  return Array.from(new Uint8Array(digest)).map((b) => b.toString(16).padStart(2, '0')).join('')
}

/** Recompute one stored line's hash. */
export async function verifyLine(line: string): Promise<{ ok: boolean; recomputed: string; stored: string }> {
  const stored = (JSON.parse(line) as Receipt).hash
  const body = bodyText(line)
  if (body === null) return { ok: false, recomputed: '', stored }
  const recomputed = await sha256Hex(body)
  return { ok: recomputed === stored, recomputed, stored }
}

/** Verify a whole ledger: every receipt's own hash, and every link to the one before it. */
export async function verifyChain(lines: string[]): Promise<ChainResult> {
  const problems: ChainResult['problems'] = []
  let prev = GENESIS
  for (let i = 0; i < lines.length; i++) {
    let r: Receipt
    try { r = JSON.parse(lines[i]) as Receipt } catch {
      problems.push({ index: i, receipt: '', problem: 'line is not valid JSON' })
      continue
    }
    const { ok, recomputed } = await verifyLine(lines[i])
    if (!ok) problems.push({ index: i, receipt: r.hash, problem: `content does not match its hash (recomputed ${recomputed.slice(0, 12)}…, stored ${String(r.hash).slice(0, 12)}…)` })
    if (r.prev_hash !== prev) problems.push({ index: i, receipt: r.hash, problem: `broken link: prev_hash ${String(r.prev_hash).slice(0, 12)}… does not follow the previous receipt ${prev.slice(0, 12)}…` })
    prev = r.hash ?? prev
  }
  return { ok: problems.length === 0, checked: lines.length, head: prev, problems }
}

export interface Ledger { lines: string[]; source: string; total: number }

/**
 * The ledgers this deployment can reach, as stored.
 *
 * There are two, and a record may sit in either: the server writes one for every capture it
 * analyses, and the build ships another covering the packs bundled with it. A static deployment has
 * only the second. Both are returned so a caller can find the receipt it is looking for rather than
 * reporting a record missing because it looked in the wrong place.
 */
export async function loadLedgers(): Promise<Ledger[]> {
  const found: Ledger[] = []
  // `no-store` on both requests, deliberately: verifying a cached copy would report a ledger intact
  // after it had been altered on disk, which is the one answer this workflow must never give.
  try {
    const token = getToken()
    const res = await fetch('/api/ledger?limit=2000', { cache: 'no-store', headers: token ? { Authorization: `Bearer ${token}` } : {} })
    if (res.ok) {
      const j = await res.json() as { lines: string[]; total: number; path: string }
      found.push({ lines: j.lines, source: j.path, total: j.total })
    }
  } catch { /* no server behind this build */ }
  try {
    const res = await fetch('/evidence/ledger.jsonl', { cache: 'no-store' })
    if (res.ok) {
      const lines = (await res.text()).split('\n').map((l) => l.trim()).filter(Boolean)
      found.push({ lines, source: 'evidence/ledger.jsonl, shipped with this build', total: lines.length })
    }
  } catch { /* nothing shipped */ }
  if (!found.length) throw new Error('No receipt ledger is reachable from this deployment.')
  return found
}

File: src/receipt.py
"""Evidence receipts: a hash chain over decisions, verifiable by someone who does not trust us.

A receipt records what was analysed (capture hash), by what (engine version, commit, acceptance
configuration), what was decided (verdict and the statistics behind it) and when. Each receipt
carries the hash of the previous one, so a receipt cannot be altered, removed or reordered without
breaking every receipt after it.

What a receipt proves, and what it does not:

- It proves the recorded decision belongs to that exact capture and that engine configuration, and
  that the ledger has not been edited since.
- It does not prove the decision is correct, and it is not a signature: anyone holding the ledger can
  recompute the chain. Signing would need a key the prototype has nowhere safe to keep.

Verification is independent of this module's own bookkeeping: `verify_chain` recomputes every hash
from the receipt content, so a receipt altered anywhere fails.
"""

import datetime
import hashlib
import json
import os
import threading

LEDGER_VERSION = 1
GENESIS = '0' * 64


def sha256_bytes(data):
    return hashlib.sha256(data).hexdigest()


def sha256_array(iq):
    """Hash of a capture as the engine saw it: interleaved float32 I/Q, little-endian."""
    import numpy as np
    a = np.asarray(iq)
    inter = np.empty(2 * a.size, dtype='<f4')
    inter[0::2] = np.real(a).astype('<f4')
    inter[1::2] = np.imag(a).astype('<f4')
    return sha256_bytes(inter.tobytes())


def canonical(obj):
    """Stable serialisation: sorted keys, no insignificant whitespace. The hash depends on content
    only, never on key order or formatting."""
    return json.dumps(obj, sort_keys=True, separators=(',', ':'), ensure_ascii=False, default=str)


def receipt_hash(body):
    return sha256_bytes(canonical(body).encode('utf-8'))


def build(*, capture_sha256, engine, decision, statistics, source=None, prev_hash=GENESIS,
          created_utc=None, capture=None):
    """Build a receipt. `engine` should carry version, commit and the acceptance configuration."""
    body = {
        'ledger_version': LEDGER_VERSION,
        'created_utc': created_utc or datetime.datetime.now(datetime.timezone.utc)
        .isoformat(timespec='seconds').replace('+00:00', 'Z'),
        'capture': {'sha256': capture_sha256, **(capture or {})},
        'engine': engine,
        'decision': decision,
        'statistics': statistics,
        'source': source,
        'prev_hash': prev_hash,
    }
    return {**body, 'hash': receipt_hash(body)}


def append(receipt, path):
    """Append a receipt to the ledger. Returns the receipt."""
    os.makedirs(os.path.dirname(path) or '.', exist_ok=True)
    with open(path, 'a', encoding='utf-8') as f:
        f.write(canonical(receipt) + '\n')
    return receipt


def last_hash(path):
    """Hash of the most recent receipt, or the genesis value when the ledger is empty."""
    if not os.path.exists(path):
        return GENESIS
    tail = GENESIS
    with open(path, encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line:
                try:
                    tail = json.loads(line).get('hash', tail)
                except json.JSONDecodeError:
                    return tail
    return tail


_CHAIN_LOCKS = {}
_CHAIN_LOCKS_GUARD = threading.Lock()


def _chain_lock(path):
    key = os.path.abspath(path)
    with _CHAIN_LOCKS_GUARD:
        lock = _CHAIN_LOCKS.get(key)
        if lock is None:
            lock = _CHAIN_LOCKS[key] = threading.Lock()
    return lock


def append_chained(path, build):
    """Read the head, build a receipt against it and append it, as one indivisible step.

    The chain invariant is that entry N's `prev_hash` is entry N-1's `hash`. Reading the head and
    appending are two separate operations, and the server is threaded: two analyses finishing at the
    same moment both read the same head, both claim it, and the ledger forks. `verify_chain` then
    reports a broken link at an entry nobody touched — a false tamper alarm, which for this system is
    worse than a missed one, because the whole point of the receipt is that a break means something.

    `build` is called with the head hash and must return the receipt to append.
    """
    # ponytail: one lock per ledger path, so one process is serialised. Two *processes* writing the
    # same ledger can still fork it; an OS file lock (msvcrt.locking / fcntl.flock) is the upgrade
    # path if the deployment ever runs more than one writer.
    with _chain_lock(path):
        receipt = build(last_hash(path))
        append(receipt, path)
        return receipt


def read_ledger(path):
    if not os.path.exists(path):
        return []
    out = []
    with open(path, encoding='utf-8') as f:
        for i, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            try:
                out.append(json.loads(line))
            except json.JSONDecodeError as e:
                raise ValueError(f'ledger line {i} is not valid JSON: {e}') from None
    return out


def verify_receipt(receipt):
    """Recompute one receipt's own hash. Returns (ok, detail)."""
    body = {k: v for k, v in receipt.items() if k != 'hash'}
    want = receipt_hash(body)
    if want != receipt.get('hash'):
        return False, (f'content does not match its hash (recomputed {want[:12]}..., '
                       f"stored {str(receipt.get('hash'))[:12]}...)")
    return True, 'hash matches content'


def verify_chain(receipts):
    """Verify a whole ledger: each receipt's hash, and each link to the one before.

    Returns a dict with ok, checked, and the problems found, naming each broken position."""
    problems = []
    prev = GENESIS
    for i, r in enumerate(receipts):
        ok, detail = verify_receipt(r)
        if not ok:
            problems.append({'index': i, 'receipt': r.get('hash'), 'problem': f'altered: {detail}'})
        if r.get('prev_hash') != prev:
            problems.append({'index': i, 'receipt': r.get('hash'),
                             'problem': (f"broken link: prev_hash {str(r.get('prev_hash'))[:12]}... "
                                         f'does not match the previous receipt {prev[:12]}...')})
        prev = r.get('hash', prev)
    return {'ok': not problems, 'checked': len(receipts), 'problems': problems,
            'head': receipts[-1]['hash'] if receipts else GENESIS}


def verify_against_capture(receipt, iq):
    """Check that a receipt belongs to this capture."""
    got = sha256_array(iq)
    want = (receipt.get('capture') or {}).get('sha256')
    return got == want, {'capture_sha256': got, 'receipt_capture_sha256': want}


def summarise(result, pack_id=None):
    """The decision and statistics a receipt records, taken from a pipeline result."""
    accept = result.get('accept') or {}
    decision = {'pack_id': pack_id, 'status': result.get('status'), 'code': result.get('code'),
                'modulation': result.get('modulation'), 'sps': result.get('sps'),
                'interleaver': result.get('interleaver'),
                'payload_bits': int(len(result.get('payload_bits', [])))}   # numpy array: len(), not truthiness
    statistics = {'log10_p': accept.get('log10_p'), 'log10_threshold': accept.get('log10_threshold'),
                  'n_hypotheses': accept.get('n_hypotheses'), 'alpha': accept.get('alpha'),
                  'families': [{k: f.get(k) for k in ('name', 'weight', 'M', 'log10_threshold',
                                                      'best_log10_p', 'accepted')}
                               for f in (accept.get('families') or [])]}
    return decision, statistics
